### ISSUE TITLE
BUG: Invalidating tracker before flush causes issue.

### ISSUE THREAD

**Issue by delgermurun** [meteorhacks/flow-router#85]
I don't know whether it is `flow-router`'s issue or `flow-layout`'s issue. Even I don't know whether it is issue or not at all.

```js
FlowRouter.route('/', {
   action: function() {
     FlowLayout.render('layout', {main: 'home'});
   }
 });
```
```js
FlowRouter.route('/posts', {
   action: function() {
     FlowLayout.render('layout', {main: 'posts'});
   }
 });
```

```js
Template.home.onCreated(function() {
   this.autorun(function() {
      console.log('created autorun', FlowRouter.reactiveCurrent());
   });
});
```

When going from `/` to `/posts`, `home` template's autorun function runs. But it shouldn't.

Cause of problem is: `flow-layout`'s reactive regions changed and immediately (before flush) tracker also invalidated. So since home template hasn't destroyed yet, `autorun` runs.

**Comment by boustanihani** [comment]
Hey, I think your issue is highly related to mine: https://github.com/meteorhacks/flow-router/issues/96

**Comment by miri-am** [comment]
@boustanihani @delgermurun. Did you find any workaround for this? I also have the problem that all my template-bound autoruns run again when a route is changed, which leads to unwanted display of information. Should we check inside the autoruns for some condition?

**Comment by elidoran** [comment]
When changing to the new route FlowLayout is only changing the ReactiveDict value for key `main` [here](https://github.com/meteorhacks/flow-layout/blob/master/lib/client/layout.js#L61).

Add this code block to see the autorun is called before the template is destroyed:

```javascript
Template.home.onRendered(function() {
    console.log('home.onRendered');
});
Template.home.onDestroyed(function() {
    console.log('home.onDestroyed');
});
```

Load '/' then go to '/posts' and we see:
```
home.onCreated autorun:  Object {path: "/", context: Context, params: Object, queryParams: Object, route: Route}
home.onRendered

home.onCreated autorun:  Object {path: "/posts", context: Context, params: Object, queryParams: Object, route: Route}
home.onDestroyed
```

So, home's autorun is rerun before the `Template.dynamic` instance reacts to the new value for `main` and destroys the home template.

This seems like the problem is inside Meteor's `Template.dynamic`, instead of the flow packages, because we need it to react before any autoruns in the templates it contains so it can destroy templates first.

**Comment by elidoran** [comment]
It seems running the reactive computations at the bottom of the hierarchy first is how Blaze does it.

Workaround? Check if it's the right path or param:

```javascript
Template.home.onCreated(function() {
   this.autorun(function() {
      var current = FlowRouter.reactiveCurrent();
      console.log('created autorun', current);
      if(current && current.path === '/') {
        // do what you want
      }
      // else it's a different path so it's about to be destroyed
   });
});
```

That requires a different style when the path is a dynamic one, such as checking the `current.params` instead.

**Comment by arunoda** [comment]
Can you check the fix-reactive-bugs branch. I did a fix for this.

**Comment by delgermurun** [comment]
@arunoda `fix-reactive-bugs` branch fixes this issue. But when you go to home page
```js
Template.home.onCreated(function() {
   this.autorun(function() {
      console.log('created autorun', FlowRouter.reactiveCurrent());
   });
});
```

`this.autorun` run twice. Even click on branch.

**Comment by arunoda** [comment]
Yep. I haven't implemented double run fix yet.

**Comment by elidoran** [comment]
With `fix-reactive-bugs` branch it doesn't run the autorun when switching routes anymore. It runs the autorun twice in a row when the route is activated though. (You both mention that above already, I'm confirming.) Still, running twice with the correct values is better than it running with the wrong value before it's destroyed.

It runs the first time when it's registered. callAction leads to FlowLayout.render leads to onCreated and the autorun registration call.

It runs the second time when line 302 is called in the afterFlush: `self._currentTracker.changed();`

**Comment by arunoda** [comment]
@elidoran can I have the source code for this case. I'll work on this.

**Comment by elidoran** [comment]
I created a basic meteor app. Deleted its starter files, added FlowLayout 1.3.0, symlinked to FlowRouter's fix-reactive-bugs branch, created the files below.

What I see when I click to switch back and forth between the Home template and the Other template: 
![screen shot 2015-05-18 at 9 08 01 pm](https://cloud.githubusercontent.com/assets/151923/7693616/20291246-fda2-11e4-913b-f4583a8029c7.png)


```coffeescript
# lib/routes.coffee
FlowRouter.route '/',
  action: -> FlowRouter.go '/app/home'

FlowRouter.route '/:layout/:template',
  action: (params) ->
    FlowLayout.render params.layout, main:params.template
```

```html
<!-- testing.html -->
<head>
  <title>Testing</title>
</head>

<body>
  <h3>Testing FlowRouter Issue #85</h3>
</body>

<template name="app">
  {{> Template.dynamic template=main}}
</template>

<template name="home">
  <h3>Home</h3>
  <div><a href="/app/other">Other</a></div>
</template>

<template name="other">
  <h3>Other</h3>
  <div><a href="/app/home">Home</a></div>
</template>
```

```coffeescript
# client/testing.coffee
Template.home.onCreated ->
  console.log 'home.onCreated'
  this.autorun (c) ->
    console.log 'autorun: first? ', c.firstRun, FlowRouter.reactiveCurrent()

Template.home.onRendered ->
  console.log 'home.onRendered'

Template.home.onDestroyed ->
  console.log 'home.onDestroyed'
```

**Comment by arunoda** [comment]
Okay. reactiveCurrent() is a different case. If you try to get getParam() it won't run two times.

**Comment by elidoran** [comment]
reactiveCurrent() is what delgermurun reported in this issue so that's what I put in the test app.

When I change to `getParam('template')` it still runs extra times:

![screen shot 2015-05-18 at 9 18 11 pm](https://cloud.githubusercontent.com/assets/151923/7693707/a0232a94-fda3-11e4-8365-cefe2f8e291d.png)

Sometimes it runs an extra time right away and other times it runs an extra time before onDestroyed is called. Sometimes the value is undefined ...

**Comment by arunoda** [comment]
Hmm. need some another look.

**Comment by arunoda** [comment]
Actually, I had a lame bug in code.
Just fixed that.
Now everything works as they should be.

I tested with your code and made a sample repo.
see: https://github.com/flow-examples/reactive-routes-example

**Comment by delgermurun** [comment]
It fixes issue.

**Comment by arunoda** [comment]
Great. I'll update test cases and do a release.

**Comment by delgermurun** [comment]
@arunoda now only `FlowRouter.reactiveCurrent()` runs twice in `autorun`. All other issues seems to be solved.

**Comment by arunoda** [comment]
Yes it. I earlier tried it with `.onRendered` then it's run only once. If we do it `onCreated` it runs twice. Lets see.

**Comment by arunoda** [comment]
Check now, this should've fixed now.

**Comment by elidoran** [comment]
I see why it still runs again before onDestroyed in my example. I'm using params in the route to control rendering so FlowRouter sees the "route" as staying the same; only its params change. The fix (which is great by the way) is based on the route changing, and mine doesn't. So this issue is resolved and what I'm seeing is a different issue most persons wouldn't encounter because they split up all their routes.

Great work solving all these reactive issues :)

**Comment by arunoda** [comment]
Thanks. We need to find a way to fix these issues anyway.

**Comment by arunoda** [comment]
Released version `1.8.0` with fixes for this bug.

**Comment by delgermurun** [comment]
:+1:
